Add Grafana Oncall support

Konstantinov-Innokentii opened the pull request.

Commit:
Add resource_amixr_integration

Commit:
Add amixr token description

julienduchesne commented on grafana/provider.go:
```diff
@@ -140,6 +141,21 @@ func Provider(version string) func() *schema.Provider {
+				"amixr_url": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					DefaultFunc:  schema.EnvDefaultFunc("GRAFANA_AMIXR_URL", "https://synthetic-monitoring-api.grafana.net"),
```
oops, probably not `https://synthetic-monitoring-api.grafana.net`

julienduchesne commented on grafana/provider.go:
```diff
@@ -140,6 +141,21 @@ func Provider(version string) func() *schema.Provider {
+					Type:         schema.TypeString,
+					Optional:     true,
+					DefaultFunc:  schema.EnvDefaultFunc("GRAFANA_AMIXR_URL", "https://synthetic-monitoring-api.grafana.net"),
+					Description:  "Synthetic monitoring backend address. May alternatively be set via the `GRAFANA_SM_URL` environment variable.",
```
Description is copied as well

julienduchesne commented on grafana/provider.go:
```diff
@@ -210,6 +242,8 @@ type client struct {
 
 	mlapi *mlapi.Client
+
+	aapi *aapi.Client
```
I'd call that `amixrAPI` even if that's a bit more verbose. I can read the others, but aapi is a bit harder

inkel replied:
I would also suggest `ocapi` or `irmapi` instead.

julienduchesne commented on grafana/provider.go:
```diff
@@ -344,6 +383,16 @@ func createSMClient(d *schema.ResourceData) (string, *smapi.Client) {
 	return smURL, smapi.NewClient(smURL, smToken, nil)
 }
 
+func createAClient(d *schema.ResourceData) (*aapi.Client, error) {
```
Same here
```suggestion
func createAmixrClient(d *schema.ResourceData) (*aapi.Client, error) {
```

julienduchesne commented on grafana/resource_amixr_integration.go:
```diff
@@ -0,0 +1,393 @@
+	}
+}
+
+func resourceIntegrationCreate(d *schema.ResourceData, m interface{}) error {
```
I think these should be prefixed as well, to be more explicit when autocompleting

```suggestion
func resourceAmixrIntegrationCreate(d *schema.ResourceData, m interface{}) error {
```

julienduchesne commented on grafana/provider.go:
```diff
@@ -164,9 +180,8 @@ func Provider(version string) func() *schema.Provider {
-				"grafana_cloud_plugin_installation": ResourceCloudPluginInstallation(),
-				"grafana_cloud_stack":               ResourceCloudStack(),
+				"grafana_cloud_api_key": ResourceCloudAPIKey(),
+				"grafana_cloud_stack":   ResourceCloudStack(),
```
Oops, removed a cloud resource there

julienduchesne reviewed (commented):
This is coming along nicely!

Commit:
Return accidentally deleted resource

Commit:
Rename aapi to amixrAPI and prefix amixr resources

Commit:
Add bunch of Amixr resources.

- Route
- Escalation Chain
- Escalation
- Schedule
- On-call shift

Commit:
Add Amixr's data sources:

- Action
- Escalation Chain
- Schedule
- Team
- User
- User Group
- SlackChanel

Commit:
Minor naming fixes

Commit:
Add handling of non-existent resources

inkel commented on grafana/adapters.go:
```diff
@@ -39,3 +51,14 @@ func int32SliceToIntList(list []int32) []interface{} {
+	var ret [][]string
+	if listSet == nil {
+		return ret
+	}
```
There's no need for this `if`: calling `range` on a `nil` slice is the same as calling it over an empty slice. What I suggest doing instead is allocating `ret` with the required size:
```suggestion
	ret := make([][]string, 0, len(listSet))
```

inkel commented on grafana/data_source_amixr_action.go:
```diff
@@ -0,0 +1,50 @@
+func DataSourceAmixrAction() *schema.Resource {
+	return &schema.Resource{
+		Description: `
+* [HTTP API](https://grafana.com/docs/grafana-cloud/oncall/oncall-api-reference/outgoing_webhooks/)
```
I believe this is going to the wrong page, maybe this is the right URL?
```suggestion
* [HTTP API](https://grafana.com/docs/grafana-cloud/oncall/oncall-api-reference/)
```

Konstantinov-Innokentii replied:
The link is correct, action and outgoing webhook are the same

Commit:
Fix handling of non-existent resources if server is down

Commit:
Add default amixr_url

Commit:
Add trailing slash to default amixr_url

Commit:
Add acceptance test for Grafana OnCall

Commit:
Add "amixr" prefix to route_test

Commit:
Linting

Commit:
Lint imports

Commit:
Fix TestProviderConfigure

Commit:
Add grafana-amixr-access-token secret

Commit:
Configute amixr client only if token provided

Commit:
Fix grafana-amixr-token secret name

Commit:
Run make drone

Commit:
Fix TestProviderConfigure

Commit:
Fix TestProviderConfigure

Commit:
Fix amixr tests

Commit:
Fix amixr tests

Commit:
Change default amixr_url

Commit:
Fix TestProviderConfigure

Commit:
Add autogenerated docs

Commit:
Fix access to amixrAPI client in tests

Commit:
Make default_route required for an integration

Commit:
Improve amixr docs, add examples

Commit:
Add import section in amixr docs

julienduchesne commented on grafana/provider.go:
```diff
@@ -344,6 +385,16 @@ func createSMClient(d *schema.ResourceData) (string, *smapi.Client) {
+		return nil, err
+	}
+	return aclient, nil
+}
```
```suggestion
	return amixrAPI.New(base_url, aToken)
}
```
You can return immediately, the client will be ignored if there's an err

julienduchesne commented on grafana/resource_amixr_route.go:
```diff
@@ -0,0 +1,169 @@
+	if client == nil {
+		err := errors.New("amixr api client is not configured")
+		return err
+	}
```
I have an idea to remove all of these blocks, but it's just sugar. We can wait until we migrate all of this to `...Context` functions

julienduchesne reviewed (approved):
I don't see any blockers to merge this. I think we can ship this and go incrementally from there!

inkel commented on grafana/data_source_amixr_escalation_chain.go:
```diff
@@ -0,0 +1,57 @@
+	client := m.(*client).amixrAPI
+	if client == nil {
+		err := errors.New("amixr api client is not configured")
+		return err
```
No need to create a new variable here:

```suggestion
		return errors.New("amixr api client is not configured")
```

inkel commented on grafana/data_source_amixr_escalation_chain.go:
```diff
@@ -0,0 +1,57 @@
+
+	if err != nil {
+		return err
+	}
```
Nit: error checking should be immediately after the action that might trigger an error is.
```suggestion
	escalationChainsResponse, _, err := client.EscalationChains.ListEscalationChains(options)
	if err != nil {
		return err
	}
```

inkel commented on grafana/data_source_amixr_schedule.go:
```diff
@@ -0,0 +1,62 @@
+	client := m.(*client).amixrAPI
+	if client == nil {
+		err := errors.New("amixr api client is not configured")
+		return err
```
```suggestion
		return errors.New("amixr api client is not configured")
```

inkel replied:
I've found the same in other files, can you please check those as well?

inkel commented on grafana/data_source_amixr_schedule.go:
```diff
@@ -0,0 +1,62 @@
+
+	if err != nil {
+		return err
+	}
```
```suggestion
	schedulesResponse, _, err := client.Schedules.ListSchedules(options)
	if err != nil {
		return err
	}
```

inkel reviewed (changes requested):
I've just left a couple suggestions regarding styling. Also, can you format the example files? I've just checked and the following aren't properly formatted.

```
$ terraform fmt -check -recursive examples
examples/provider/provider-amixr.tf
examples/resources/grafana_amixr_escalation/resource.tf
examples/resources/grafana_amixr_on_call_shift/grafana_amixr_on_call_shift.tf
examples/resources/grafana_amixr_route/resource.tf
examples/resources/grafana_amixr_schedule/resource.tf
```

Commit:
Rename amixr -> oncall

Commit:
Format examples

Commit:
Recreate oncall docs

Commit:
Style fixes

Commit:
Rename amixr to oncall in drone

Commit:
Fix exmaple for grafana_oncall_outgoing_webhook

Commit:
Lint

Commit:
Fix TestProviderConfigure

Commit:
format errors

Commit:
Rename outgoing_webhook back to action

Commit:
Add docs oncall_action

Commit:
Fix docs drift

inkel commented on grafana/provider_test.go:
```diff
@@ -85,7 +85,7 @@ func TestProviderConfigure(t *testing.T) {
 			name:        "no config",
 			env:         map[string]string{},
-			expectedErr: "\"auth\": one of `auth,cloud_api_key,sm_access_token` must be specified",
+			expectedErr: "\"auth\": one of `auth,cloud_api_key,oncall_access_token,sm_access_token` must\nbe specified",
```
I'm curious: where does this `\n` between `must` and `be` come from? Checking the [SDK code](https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L1874) that raises this error doesn't seem to have a newline, maybe it's the `resource.UnitTest`? 🤔

inkel reviewed (approved):
Excellent work, @Konstantinov-Innokentii!!! Thanks for your patience with all the change requests ❤️ 

Let's wait for @julienduchesne for a final approval.
